### PR TITLE
EZEE-3176: Reduced scope of content changes to new and updated fields

### PR DESCRIPTION
**Pull request by Nattfarinn**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [EZEE-3176](https://jira.ez.no/browse/EZEE-3176)
| **Type**                                   | improvement
| **Target eZ Platform version** | `v3.1`
| **BC breaks**                          | no
| **Tests pass**                          | yes
| **Doc needed**                       | no

#### Checklist:
- [x] PR description is updated.
- [x] Tests are implemented.
- [x] Added code follows Coding Standards (use `$ composer fix-cs`).
- [x] PR is ready for a review.

### CODE REVIEW

File: eZ/Publish/Core/Repository/ContentService.php [hunk 1]
```diff
@@ -580,6 +580,8 @@ public function createContent(APIContentCreateStruct $contentCreateStruct, array
             $contentCreateStruct->contentType->id
         );
 
+        $contentCreateStruct->fields = $this->contentMapper->getFieldsForCreate($contentCreateStruct->fields, $contentCreateStruct->contentType);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentService.php:583]: CS: this line is too long

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 2]
```diff
@@ -159,6 +167,42 @@ public function mapFieldsForUpdate(
         return $fields;
     }
 
+    public function getFieldValueForCreate(FieldDefinition $fieldDefinition, ?Field $field): Value
+    {
+        if (isset($field)) {
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:172]: This is not strict, did you mean:
```suggestion
        if (null !== $field) {
```
?

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 3]
```diff
@@ -159,6 +167,42 @@ public function mapFieldsForUpdate(
...
+
+        $fieldType = $this->fieldTypeRegistry->getFieldType(
+            $fieldDefinition->fieldTypeIdentifier
+        );
+
+        return $fieldType->acceptValue($fieldValue);
+    }
+
+    public function getFieldValueForUpdate(?Field $newField, ?Field $previousField, FieldDefinition $fieldDefinition, bool $isLanguageNew): Value
+    {
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:186]: CS: please reformat this, lines are too long.

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 4]
```diff
@@ -159,6 +167,42 @@ public function mapFieldsForUpdate(
...
+        $isFieldUpdated = isset($newField);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:187]:
```suggestion
        $isFieldUpdated = null !== $newField;
```
?

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 5]
```diff
@@ -227,4 +271,80 @@ private function cloneField(Field $field, array $overrides = []): Field
...
+            }
+
+            $fieldType = $this->fieldTypeRegistry->getFieldType($fieldDefinition->fieldTypeIdentifier);
+
+            $field = $content->getField($updatedField->fieldDefIdentifier);
+            $updatedFieldValue = $this->getFieldValueForUpdate(
+                $updatedField,
+                $field,
+                $contentType->getFieldDefinition($updatedField->fieldDefIdentifier),
+                !in_array($updatedField->languageCode, $content->versionInfo->languageCodes)
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:302]:
```suggestion
                !in_array($updatedField->languageCode, $content->versionInfo->languageCodes, true)
```

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 6]
```diff
@@ -227,4 +271,80 @@ private function cloneField(Field $field, array $overrides = []): Field
...
+            );
+
+            if (!empty($field)) {
+                $updatedFieldHash = md5(serialize($fieldType->toHash($updatedFieldValue)));
+                $contentFieldHash = md5(serialize($fieldType->toHash($field->value)));
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:307]: can we avoid `serialize` in favor of e.g. `json_encode` or something else? `serialize` proved to have security vulnerabilities in the past.

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 7]
```diff
@@ -227,4 +271,80 @@ private function cloneField(Field $field, array $overrides = []): Field
...
+        }
+
+        return $fields;
+    }
+
+    public function getFieldsForCreate(array $createdFields, ContentType $contentType): array
+    {
+        $fields = [];
+
+        /** @var Field $createdField */
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:322]: what kind of Field is that API or SPI? Please use FQCN ;)
Side: if you'd document method arguments, would it still be required?

File: eZ/Publish/Core/Repository/Mapper/ContentMapper.php [hunk 8]
```diff
@@ -227,4 +271,80 @@ private function cloneField(Field $field, array $overrides = []): Field
...
+
+            $fieldType = $this->fieldTypeRegistry->getFieldType($fieldDefinition->fieldTypeIdentifier);
+
+            $createdFieldValue = $this->getFieldValueForCreate(
+                $fieldDefinition,
+                $createdField
+            );
+
+            $createdFieldHash = md5(serialize($fieldType->toHash($createdFieldValue)));
+            $defaultFieldHash = md5(serialize($fieldType->toHash($fieldDefinition->defaultValue)));
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Mapper/ContentMapper.php:341]: the same comment regarding serialization.

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php [hunk 9]
```diff
@@ -124,6 +134,11 @@ function (Field $field) {
             array_unique($languageCodes)
         );
 
+        if (!empty($this->content)) {
+            dump($fields);
+            dump($this->content);
+        }
+
```
Comment (alongosz) [inline eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php:141]: This is not gonna work in prod :P

File: eZ/Publish/Core/Repository/Validator/VersionValidator.php [hunk 10]
```diff
@@ -43,10 +44,10 @@ public function validate(
             throw new InvalidArgumentException('$object', 'Not supported');
         }
 
-        /** @var \eZ\Publish\API\Repository\Values\Content\VersionInfo $versionInfo */
-        $versionInfo = $object;
+        if (empty($context['content']) || !$context['content'] instanceof Content) {
+            throw new InvalidArgumentException('$context', 'Context "content" parameter is not an instance of ' . Content::class);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Validator/VersionValidator.php:48]: CS: this line is too long, maybe `\sprintf`?

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php [hunk 11]
```diff
@@ -23,11 +24,20 @@ final class VersionBuilder
...
     private $targetVersionProperties = [];
 
+    private $updatedFields = [];
+
     public function build(): Target\Version
     {
         return new Target\Version($this->targetVersionProperties);
     }
 
+    public function updateFields(array $updatedFields): self
```
Comment (alongosz) [inline eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php:34]: are those instances of `Field` or are they actually `$updatedFieldIdentifiers`?

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php (hunk 9 above)
Comment (Nattfarinn) [reply eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php:141]: Oops. Squashed dev commit by accident :P

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php (hunk 11 above)
Comment (Nattfarinn) [reply eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php:35]: `\eZ\Publish\API\Repository\Values\Content\Field[]`, docblock added.

**Pull request merged**